File: 228-summary-ranges/solution.cpp
#include <vector>
#include <string>
#include <sstream>

class Solution
{
private:
  static std::string make_range_str(const int start, const int end)
  {
    std::ostringstream oss;
    if (start == end)
    {
      oss << start;
    }
    else
    {
      oss << start << "->" << end;
    }

    return oss.str();
  }

public:
  std::vector<std::string> summaryRanges(std::vector<int> &nums)
  {
    std::vector<std::string> result;
    if (nums.empty())
    {
      return result;
    }

    int start = nums[0], end = nums[0];
    for (int i = 1; i < nums.size(); ++i)
    {
      const auto num = nums[i];
      if (num == end + 1)
      {
        end = num;
        continue;
      }

      result.emplace_back(Solution::make_range_str(start, end));
      start = end = num;
    }

    result.emplace_back(Solution::make_range_str(start, end));
    return result;
  }
};

File: 228-summary-ranges/solution.py
class Solution:
  def summaryRanges(self, nums: list[int]) -> list[str]:
    result: list[str] = []
    if not nums:
      return result

    start = end = None
    for num in nums:
      if start is None or end is None:
        start = end = num
        continue

      if num == end + 1:
        end = num
        continue

      result.append(str(start) if start == end else f'{start}->{end}')
      start = end = num

    result.append(str(start) if start == end else f'{start}->{end}')
    return result
